### DISCUSSION TITLE
バチバチなボール回り込み

### DISCUSSION THREAD

**Discussion by HansRobo** [ibis-ssl/crane#414] [General] [answered]
ボールに追いついた瞬間にはパスが出ているくらいの魔法のようなバッチバチのボール回り込みを実現したい。
この過程で必要な修正は、ビジョンフィルタリング・ビヘイビア制御・経路生成・AIコマンド・ローカル制御など多岐かつ大幅な変更を必要とするかも知れないがそれでも実現したい。（学ロボを見て完全に影響を受けている）

**Comment by HansRobo** [comment]
## カーペットモデル

回り込みにはボールの軌道を正確に予測するためのカーペットモデルが必要

**Reply by ChikuwaRobo** [reply]
JOで中国チームがボールの転がりを確認していた気がする。

**Reply by HansRobo** [reply]
モデルは作れても係数は結局実際のフィールドで計測することになりそうね

**Reply by HansRobo** [reply]
ボールの回転数を状態変数に入れ込めれば、カーブシュートやバックスピンなどに対応したモデルを作ることが出来るかも

**Reply by HansRobo** [reply]
現在は、等加速度直線運動モデルを使っている。
カーペットの摩擦による減速を加速度（減速度）に反映して調整する。
https://github.com/ibis-ssl/crane/blob/cd015fe0257ce65ef7a9754fbd6793979255e7a8/utility/crane_basics/include/crane_basics/ball_model.hpp#L22

**Comment by HansRobo** [comment]
## ビジョンフィルタリング

ビジョンの遅れを陽に扱い、フィールド状況とAIの把握状況の差を最小限に抑える

**Comment by HansRobo** [comment]
## 軌道生成

回り込み軌道を以下の観点から考える

- 回り込み完了時間の短さ
- 成功確率の高さ

**Reply by ChikuwaRobo** [reply]
距離的に大きく回り込むことによる時間的ペナルティはわりと小さいはず。(移動距離は速度と加速度で殴ってどうにかなるし)
成功率や、回り込んだあとどうするかの自由度が高いので、こぼれたボールを拾いに行くような場合は、大きめに回り込むほうがいいかも

**Reply by HansRobo** [reply]
ローカルカメラのことを考えるなら、いかに早くローカルカメラの視界にボールを入れてローカル制御に引き継ぎやすいところに持っていくかも重要そう。

**Comment by ChikuwaRobo** [comment]
## ローカルカメラによる制御
当初はキーパーのためにローカルカメラを搭載していたが、近距離でのボールハンドリングのほうがローカルカメラ適性が高いかもしれない。
特に一瞬トラップしてからシュートへ移るような動きには効果的なのではないかと思う

**Reply by ChikuwaRobo** [reply]
特にボールに回転がかかっている状態では、予測との不一致が大きくなる=遅延の悪影響が強い
ため、低レイテンシなセンサーが必要
(非接触でボールの回転を取得できれば最強なんだが)

**Reply by ChikuwaRobo** [reply]
ボールへのアプローチをロボット側で動かすなら、
座標推定→軌道予測→到達可能ポイントへ追従よりも、
比例航法とか使ったほうが良さそう
https://rikei-tawamure.com/entry/2020/03/20/192853

**Reply by HansRobo** [reply]
一人称視点なら比例航法はかなりありだね。
実装が簡単な割に実用性が高そう

**Reply by HansRobo** [reply]
ローカルカメラのレイテンシを改善するモチベが出てきたな。
というか、ローカルキーパーもこの比例航法を使ったボールアプローチをそのまま使えそう。

**Reply by ChikuwaRobo** [reply]
RasPi OSも現状普通のイメージ使ってるのでRTカーネルにしたほうが良いかも

**Reply by HansRobo** [reply]
GreenTeaが既にローカルカメラOnlyのボールを扱うモードを実装していた。
やはり、専用人材がいるのは強い。が、AI側が対応できていないらしく試合では使えないらしい

**Comment by HansRobo** [comment]
## RoboCupジュニアの知見の取り込み

これそのまま実装すれば良さそう
https://yuta.techblog.jp/archives/40889399.html

**Reply by HansRobo** [reply, marked as answer]
> ボールに追いついた瞬間にはパスが出ている

これは↑の実装によって実現された。あとは速度が遅いなどの細かな調整あるのみ
